docs: migrate sphinx api docs to mkdocs

ccmao1130 opened the pull request:
## Part 1
- [x] Dataframe Creation
- [x] Dataframe
- [x] Expressions (namespaces: url, float, datetime, string, list, struct, map, projection, image, partitioning, json, embedding, binary)
- [x] Functions
- [ ] SQL (all sql funcs)
- [x] Sessions
- [x] Catalogs
- [x] Tables
- [x] Schema
- [x] Data types
- [x] Groupby
- [x] UDFs
- [x] Series
- [x] Configuration (s3config, s3credentials, gcsconfig, azureconfig)
- [x] Miscellaneous

## Part 2
- [x] rename `mkdocs` folder to `user-guide` to differentiate between user guide and api docs
     ^ update: all user guide docs will live in `docs` and all api docs will live in `docs/api_docs`, this is so that build site opens on user guide
- [x] update relative links between user guide and api docs
     - [x] Terminology 
     - [x] Quickstart
     - [x] Core Concepts
     - [x] SQL Reference
     - [x] Sessions
     - [x] Catalogs
     - [x] Spark Connect
     - [x] Distributed Computing
     - [x] Advanced
     - [x] Integrations
          - [x] Ray
          - [x] Unity
          - [x] Apache Iceberg
          - [x] Delta Lake
          - [x] Apache Hudi
          - [x] AWS (blocked by s3config)
          - [x] Azure (blocked by azureconfig)
          - [x] SQL
          - [x] Hugging Face Datasets
          - [x] S3 Tables
     - [x] Resources
     - [x] Migration Guide

## Part 3
- [ ] reformat docstrings and examples for api docs https://github.com/Eventual-Inc/Daft/pull/4148
    - [ ] fix warning/note boxes from original docstring
- [ ] fix random headings that shouldn't be
- [ ] how can we better group the integrations section?

## Part 4 / Clean Up (after complete migration is done)
- [x] remove from mkdocs.yml
```
- mkdocs-simple-hooks:
    hooks:
      on_pre_build: "docs.hooks:make_api_docs"
```
- [x] remove `docs/hooks.py` file
- [ ] delete `docs/Makefile` (these are sphinx build instructions)
- [ ] remove sphinx folders in `docs/.gitignore` and only keep `site`
- [ ] remove sphinx mentions in `.github/workflows/build-docs.yml`
- [x] update `Makefile` with `make docs` to build docs and `make serve` to build in development serve
- [x] update `docs/CONTRIBUTING.md` with build command `make docs` and serve command `make serve` in `Daft` dir not in `docs` dir
- [ ] search `todo(docs - cc)` for missing items
- [ ] remove docs/source/... in `.pre-commit-config.yml`

ccmao1130 commented:
here's the latest version: https://www.getdaft.io/projects/docs/en/chanchan-api-docs/

- i haven't removed all of the sphinx stuff yet, was planning on doing that in another pr, this builds successfully without any mention of sphinx
- the docstring examples still look weird in this version but i have a separate pr to reformat all of the docstrings from rst to md format (https://github.com/Eventual-Inc/Daft/pull/4148)

jaychia commented:
Looks good overall, going to put a couple of thoughts here:

I think `sql_expr` qualifies as a Constructor
<img width="208" alt="image" src="https://github.com/user-attachments/assets/19e54aad-2c72-48df-ad6b-b92d2f1e54c8" />

`_expr` should not be publicly documented
<img width="174" alt="image" src="https://github.com/user-attachments/assets/8c4712d9-1de5-418f-bca5-1aaa49d5bc15" />

We should make Projection private. I see that we refer to this in certain parts of our public facing API (e.g. groupby) but we should just simplify there to change those to simply list[Expression] I think.
<img width="559" alt="image" src="https://github.com/user-attachments/assets/9585af32-785a-4a89-a112-0b153f85c352" />
<img width="227" alt="image" src="https://github.com/user-attachments/assets/79908c00-9102-4619-b503-10fe4253dd4b" />

Could be nice to explicitly call out Dataframe methods that will execute compute (i.e. `collect()`, `write_*`, `show()`, `to_*`...)
<img width="221" alt="image" src="https://github.com/user-attachments/assets/adf3f43e-2509-4fc1-af2e-c71e7585b9b4" />

Functions vs Expressions is a little confusing atm. They seem to be two concepts that are very much the same? We should talk about introducing new concepts.

For `Series`, I think we should be careful as to what we want users to use. I'm in favor of not exposing all the methods available on it, only the "conversion" methods to/from other formats such as arrow and numpy.
<img width="229" alt="image" src="https://github.com/user-attachments/assets/9ed4ba39-8208-4e9f-b055-807e55fc994a" />

Overall great work! In favor of merging this in first and working on these incrementally depending on what we deem as high priority.

ccmao1130 commented:
@jaychia thanks for your review!

Done:
- moved sql_expr to constructors
- removed Projections
- limit Series to conversion to/from methods

Next Iteration:
- explicitly call out Dataframe methods that will execute compute (i.e. collect(), write_*, show(), to_*...)
- “Functions vs Expressions is a little confusing atm. They seem to be two concepts that are very much the same? We should talk about introducing new concepts.”

rchowell reviewed (approved):
The SQL pages look nice, thanks for moving them. I will need to add more content here though because they are a bit sparse.

universalmind303 commented on daft/convert.py:
```diff
@@ -84,12 +89,22 @@ def from_pydict(data: Dict[str, InputListType]) -> "DataFrame":
+    Returns:
+        DataFrame: DataFrame created from the provided pyarrow Table.
+
     Example:
```
change to `Examples:` and use doctest style formatting

universalmind303 commented on daft/convert.py:
```diff
@@ -154,11 +169,11 @@ def from_pandas(data: Union["pd.DataFrame", List["pd.DataFrame"]]) -> "DataFrame
     Args:
         ds: The Ray Dataset to create a Daft DataFrame from.
+
+    !!! note "This function can only work if Daft is running using the RayRunner."
```
use `Note:` instead of manual admonition

universalmind303 commented on daft/convert.py:
```diff
@@ -169,14 +184,11 @@ def from_ray_dataset(ds: "RayDataset") -> "DataFrame":
     Args:
         ddf: The Dask DataFrame to create a Daft DataFrame from.
+
+    !!! note "The provided Dask DataFrame must have been created using [Dask-on-Ray](https://docs.ray.io/en/latest/ray-more-libs/dask-on-ray.html). This function can only work if Daft is running using the RayRunner."
```
use `Note:` instead of manual admonition

universalmind303 commented on daft/dataframe/dataframe.py:
```diff
@@ -94,17 +94,14 @@ def _utc_now() -> datetime:
-    It has columns, where each column has a type and the same
-    number of items (rows) as all other columns.
-    """
+    # """A Daft DataFrame is a table of data.
+    # It has columns, where each column has a type and the same number of items (rows) as all other columns.
+    # """
```
Uncomment

universalmind303 commented on daft/dataframe/dataframe.py:
```diff
@@ -291,14 +288,18 @@ def iter_rows(
+                available on the machine).
+            column_format: the format of the columns to iterate over. One of "python" or "arrow". Defaults to "python".
+
+        !!! note "A quick note on configuring asynchronous/parallel execution using `results_buffer_size`."
```
use `Note:` instead of manual admonition

universalmind303 commented on daft/dataframe/dataframe.py:
````diff
@@ -310,23 +311,20 @@ def iter_rows(
-            ...     print(row)
-            {'foo': 1, 'bar': 'a'}
-            {'foo': 2, 'bar': 'b'}
-            {'foo': 3, 'bar': 'c'}
+            ``` py linenums="1"
+            import daft
 
+            df = daft.from_pydict({"foo": [1, 2, 3], "bar": ["a", "b", "c"]})
+            for row in df.iter_rows():
+                print(row)
+            ```
+            ```
+            {"foo": 1, "bar": "a"}
+            {"foo": 2, "bar": "b"}
+            {"foo": 3, "bar": "c"}
+            ```
````
change to `Examples:` and use doctest style formatting

universalmind303 reviewed (changes requested):
It looks like a lot of the docstrings are incorrectly formatted still. 
Some are using markdown, others are still using the doctest style. 

There's also an inconsistency between `Example:` and `Examples:`. 

Also need to remove the admonitions: 
```
!!!<label> "<text>"
``` 

and use mkdocstrings to generate the admonitions: 

```
<label>:
  <text>
```

universalmind303 commented on Makefile:
```diff
@@ -76,12 +76,12 @@ dsdgen: .venv ## Generate TPC-DS data
-.PHONY: sphinx-docs
-sphinx-docs: .venv ## Build Sphinx API documentation
-	uv run --with-requirements requirements-docs.txt sphinx-build -b html "docs/sphinx/source" "docs/sphinx/_build"
+.PHONY: serve
+serve: .venv ## Build Daft documentation in development server
```
rename to `docs-serve`
